src/main/java/com/wanderersoftherift/wotr/block/blockentity/AnomalyBlockEntity.java

```diff
@@ -0,0 +1,230 @@
+import com.wanderersoftherift.wotr.init.WotrBlockEntities;
+import com.wanderersoftherift.wotr.serialization.DispatchedPairOptionalValue;
+import com.wanderersoftherift.wotr.util.RandomSourceFromJavaRandom;
+import net.minecraft.client.multiplayer.ClientLevel;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.core.particles.DustParticleOptions;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ClientGamePacketListener;
+import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+
+public class AnomalyBlockEntity extends BlockEntity {
+
+    private static final float COMPLETED_STATE = 0.1f;
+    private static final float INCOMPLETE_STATE = 1f;
+    private long seed = 0L;
+    private Holder<AnomalyReward> reward;
+    private AnomalyState<?> state;
+
+    public AnomalyBlockEntity(BlockPos pos, BlockState state) {
+        super(WotrBlockEntities.ANOMALY_BLOCK_ENTITY.get(), pos, state);
+    }
+
+    public void clientTick(ClientLevel clientLevel, BlockPos pos, BlockState state1) {
+        if (state != null) {
+            double centerX = pos.getX() + 0.5;
+            double centerY = pos.getY();
+            double centerZ = pos.getZ() + 0.5;
+
+            var count = 6;
+            for (int i = 0; i < count; i++) {
+                double angle = (level.getGameTime() - 25.0 * Math.sin(0.03 * level.getGameTime() + 1.2)) * 0.2
+                        + i * Math.PI * 2 / count;
+                double radius = 0.4 + Math.sin(level.getGameTime() * 0.03) * 0.25;
+
+                double x = centerX + Math.cos(angle) * radius;
+                double z = centerZ + Math.sin(angle) * radius;
+                double y = centerY + 0.5 + Math.sin(level.getGameTime() * 0.5 + i * Math.PI * 2 / count) * 0.1
+                        * (1 + Math.sin(0.03 * level.getGameTime()));
+
+                var particleColor = state.task.value().particleColor();
+
+                level.addParticle(new DustParticleOptions(particleColor, 0.4f), true, false, x, y, z, 0.0, 0.0, 0.0);
+            }
+        }
+    }
+
+    public void scheduledTick(ServerLevel serverLevel, BlockPos pos, BlockState state1) {
+        if (state == null) {
+            return;
+        }
+        var taskState = state.state();
+        if (taskState.isPresent() && taskState.get() instanceof BattleTaskState battleTask) {
```

Comment:
Feels odd to have BattleTaskState behavior hardcoded here, maybe should have a tick method in taskState and delegate to that.